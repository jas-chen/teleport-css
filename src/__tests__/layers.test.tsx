import { expect, test } from 'vitest';
import { renderToStaticMarkup } from 'react-dom/server';
import fnv1a from '@sindresorhus/fnv1a';
import { create } from '..';
import pretty from 'pretty';

function hashFn(value: string) {
  return fnv1a(value, { size: 64 }).toString(36).slice(0, 8);
}

const { styled } = create({
  hashFn,
  defaultLayer: 'utilities',
});

test('can set default layer', () => {
  const Button = styled('button', () => ({
    color: 'red',
    background: 'blue',
  }));

  expect(pretty(renderToStaticMarkup(<Button />))).toMatchSnapshot();
});

test('CSS in layers should not be atomic', () => {
  const Button = styled('button', () => ({
    color: 'red',
    '@layer component': {
      opacity: 1,
      position: 'relative',
      '&:hover': {
        color: 'blue',
        opacity: 0.5,
      },
    },
  }));

  expect(pretty(renderToStaticMarkup(<Button />))).toMatchSnapshot();
});

test('CSS layers composition - 1', () => {
  const ButtonA = styled('button', () => ({
    '@layer component': {
      opacity: 1,
      position: 'relative',
    },
  }));

  const ButtonB = styled('button', () => ({
    '@layer component': {
      opacity: 0.5,
      position: 'static',
    },
  }));

  const ButtonC = styled(ButtonB, () => ({
    '@layer component': {
      opacity: 1,
      position: 'relative',
    },
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <ButtonA />
          <ButtonC />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});

test('CSS layers composition - 2', () => {
  const ButtonA = styled('button', () => ({
    '@layer component.test': {
      opacity: 1,
      position: 'relative',
    },
  }));

  const ButtonB = styled('button', () => ({
    '@layer component': {
      opacity: 0.5,
      position: 'static',
    },
  }));

  const ButtonC = styled(ButtonB, () => ({
    '@layer component.test': {
      opacity: 1,
      position: 'relative',
    },
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <ButtonA />
          <ButtonC />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});
