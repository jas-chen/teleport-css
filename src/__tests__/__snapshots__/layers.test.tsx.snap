// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSS in layers should not be atomic 1`] = `
"<style data-precedence="1" data-href="x_4tcsx9ze1 x_3eivjhwl1">
  .x_4tcsx9ze1 {
    @layer utilities {
      color: red
    }
  }

  .x_3eivjhwl1 {
    @layer component {
      opacity: 1;
      position: relative;

      &:hover {
        color: blue;
        opacity: 0.5
      }
    }
  }
</style><button class="x_4tcsx9ze1 x_3eivjhwl1"></button>"
`;

exports[`CSS layers composition - 1 1`] = `
"<style data-precedence="1" data-href="x_2dj17gws1 x_2c1ksewg1">
  .x_2dj17gws1 {
    @layer component {
      opacity: 1;
      position: relative
    }
  }

  .x_2c1ksewg1 {
    @layer component {
      opacity: 0.5;
      position: static
    }
  }
</style>
<style data-precedence="2" data-href="x_2dj17gws2">
  .x_2dj17gws2 {
    @layer component {
      opacity: 1;
      position: relative
    }
  }
</style>
<div><button class="x_2dj17gws1"></button><button class="x_2c1ksewg1 x_2dj17gws2"></button></div>"
`;

exports[`CSS layers composition - 2 1`] = `
"<style data-precedence="1" data-href="x_2fpp4j2h1 x_2c1ksewg1">
  .x_2fpp4j2h1 {
    @layer component.test {
      opacity: 1;
      position: relative
    }
  }

  .x_2c1ksewg1 {
    @layer component {
      opacity: 0.5;
      position: static
    }
  }
</style>
<style data-precedence="2" data-href="x_2fpp4j2h2">
  .x_2fpp4j2h2 {
    @layer component.test {
      opacity: 1;
      position: relative
    }
  }
</style>
<div><button class="x_2fpp4j2h1"></button><button class="x_2c1ksewg1 x_2fpp4j2h2"></button></div>"
`;

exports[`can set default layer 1`] = `
"<style data-precedence="1" data-href="x_4tcsx9ze1 x_1fwwk7jh1">
  .x_4tcsx9ze1 {
    @layer utilities {
      color: red
    }
  }

  .x_1fwwk7jh1 {
    @layer utilities {
      background: blue
    }
  }
</style><button class="x_4tcsx9ze1 x_1fwwk7jh1"></button>"
`;
